### PR TITLE
Implement sync for multi ig node pool

### PR DESCRIPTION
**Pull request by panslava**

### CODE REVIEW

File: pkg/instancegroup/multiigcontroller.go [hunk 1]
```diff
@@ -88,11 +97,100 @@ func (igc *MultiIGController) Shutdown() {
...
 
+func (igc *MultiIGController) splitNodesByIg(nodeToIg map[string]*compute.InstanceGroup) map[string][]string {
+	igToNodes := make(map[string][]string)
+	for node, ig := range nodeToIg {
+		igToNodes[ig.Name] = append(igToNodes[ig.Name], node)
+	}
+	return igToNodes
+}
+
+func (igc *MultiIGController) getSuffixedIgNameByIndex(index int) string {
```
Comment (kl52752) [inline pkg/instancegroup/multiigcontroller.go:108]: I think that getIgNameByIndex is better name

File: pkg/instancegroup/multiigcontroller.go [hunk 2]
```diff
@@ -88,11 +97,100 @@ func (igc *MultiIGController) Shutdown() {
...
+	for igIndex := 0; len(addNodes) != 0; igIndex++ {
+		currentIgName := igc.getSuffixedIgNameByIndex(igIndex)
+		if _, exist := igSizes[currentIgName]; !exist {
+			err := igc.instancePool.Cloud.CreateInstanceGroup(&compute.InstanceGroup{Name: currentIgName}, zone)
+			if err != nil {
+				return err
+			}
+			igSizes[currentIgName] = 0
+		}
+		availableSize := maxIgSize - igSizes[currentIgName]
```
Comment (kl52752) [inline pkg/instancegroup/multiigcontroller.go:189]: you can just use F.MaxIgSize it is exported

**Review by kl52752 — CHANGES_REQUESTED** [review]

File: pkg/instancegroup/multiigcontroller.go (hunk 1 above)
Comment (kl52752) [reply pkg/instancegroup/multiigcontroller.go:108]: What do you think of creating new namer just like for L4? I don't think that this function belongs to controller.

Comment (panslava) [reply pkg/instancegroup/multiigcontroller.go]: Added function to an old namer

File: pkg/instancegroup/multiigcontroller.go (hunk 2 above)
Comment (panslava) [reply pkg/instancegroup/multiigcontroller.go]: Done

File: pkg/utils/namer/namer.go [hunk 3]
```diff
@@ -323,6 +323,16 @@ func (n *Namer) InstanceGroup() string {
 	return n.decorateName(n.prefix + "-" + igPrefix)
 }
 
+// InstanceGroupByIndex constructs the name for an Instance Group when multiple Instance Groups are in use.
+func (n *Namer) InstanceGroupByIndex(index int) string {
```
Comment (kl52752) [inline pkg/utils/namer/namer.go:327]: Can you extract it to separate PR (with unit tests) ?

File: pkg/utils/namer/namer.go [hunk 4]
```diff
@@ -323,6 +323,16 @@ func (n *Namer) InstanceGroup() string {
...
+	name := n.InstanceGroup()
+	if index == 0 {
+		return name
+	} else {
```
Comment (kl52752) [inline pkg/utils/namer/namer.go:331]: else if not necessary just return value

Comment (panslava) [reply pkg/utils/namer/namer.go]: fixed

File: pkg/instancegroup/multiigcontroller.go [hunk 5]
```diff
@@ -88,11 +95,99 @@ func (igc *MultiIGController) Shutdown() {
...
+		klog.V(2).Infof("Removing %d, adding %d nodes", len(nodesToRemove), len(nodesToAdd))
+
+		if err = igc.removeNodes(nodesToRemove, zone, nodesToIG); err != nil {
+			return err
+		}
+		if err = igc.addNodes(nodesToAdd, zone, igSizes); err != nil {
+			return err
+		}
+	}
+	return igc.instancePool.Sync(allNodeNames)
```
Comment (cezarygerard) [inline pkg/instancegroup/multiigcontroller.go:153]: why calling this?

the
"func (igc *MultiIGController) sync(key string) error "

replaces instancePool.Sync

File: pkg/instancegroup/multiigcontroller.go [hunk 6]
```diff
@@ -88,11 +95,99 @@ func (igc *MultiIGController) Shutdown() {
...
+	igToNodes := make(map[string][]string)
+	for node, ig := range nodeToIG {
+		igToNodes[ig.Name] = append(igToNodes[ig.Name], node)
+	}
+	return igToNodes
+}
+
 func (igc *MultiIGController) sync(key string) error {
-	_, err := utils.GetReadyNodeNames(listers.NewNodeLister(igc.lister))
+	allNodeNames, err := utils.GetReadyNodeNames(listers.NewNodeLister(igc.lister))
```
Comment (cezarygerard) [inline pkg/instancegroup/multiigcontroller.go:107]: NodeLister should be a field in MultiIGController
instead of igc.lister

**Pull request closed without merging**